### PR TITLE
CORDA-540: Implementation of path in the serialization graph

### PR DESCRIPTION
**Pull request by vkolomeyko**

### CODE REVIEW

**Comment by rick-r3** [comment]
I think it is better to implement this as something that isn't in the code path _until_ there's an exception.  Your current implementation maintains a stack manually, but there's an actual program stack that will unwind when there's an exception, and you can leverage that to make it less burdensome to the happy path when there are no exceptions.

If each code block has a `try/catch` and then in `catch` prepends the exception message with the bit of path that is pertinent to the current serializer (class name if ObjectSerializer, property name if a PropertySerializer) and then re-throws it will provide the same utility but not be executed except when needed.

If you make the `try/catch` part of an inline lambda, then it starts to become very efficient both in source code and execution.

**Comment by vkolomeyko** [comment]
Sure - will re-work the implementation.

**Comment by vkolomeyko** [comment]
Re-worked as requested.
Example exceptions will look as follows:
```
java.io.NotSerializableException: outputs(java.util.List<net.corda.core.contracts.TransactionState<net.corda.core.contracts.ContractState>>) -> java.util.List<net.corda.core.contracts.TransactionState<net.corda.core.contracts.ContractState>> -> constraint(net.corda.core.contracts.AttachmentConstraint) -> Class interface net.corda.core.contracts.AttachmentConstraint is not on the whitelist or annotated with @CordaSerializable.

	at net.corda.nodeapi.internal.serialization.amqp.SerializerFactory.whitelisted(SerializerFactory.kt:274)
	at net.corda.nodeapi.internal.serialization.amqp.SerializerFactory.access$whitelisted(SerializerFactory.kt:33)
	at net.corda.nodeapi.internal.serialization.amqp.SerializerFactory$makeClassSerializer$1.apply(SerializerFactory.kt:245)
	at net.corda.nodeapi.internal.serialization.amqp.SerializerFactory$makeClassSerializer$1.apply(SerializerFactory.kt:33)
```

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt [hunk 1]
```diff
@@ -237,4 +237,27 @@ internal fun suitableForObjectReference(type: Type): Boolean {
...
+/**
+ * Utility function which helps tracking the path in the object graph when exceptions are thrown.
+ * Since there might be a chain of nested calls it is useful to record which part of teh graph caused an issue.
+ * Path information is added to the message of teh exception being thrown.
+ */
+internal inline fun <T> ifThrowsAppend(strToAppendFn: () -> String, block: () -> T): T {
+    try {
+        return block()
+    } catch (th: Throwable) {
+        th.setMessage("${strToAppendFn()} -> " + th.message)
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: you can make the `+ th.message` also use the $ syntax

**Pull request merged**